plugins/xdebug/xdebug_headers.cc

```diff
@@ -0,0 +1,159 @@
+/** @file
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include <cstdlib>
+#include <stdio.h>
+#include <cstdio>
+#include <strings.h>
+#include <string_view>
+#include <sstream>
+#include <cstring>
+#include <getopt.h>
+
+#define DEBUG_TAG_LOG_HEADERS "xdebug.headers"
+
+std::string_view
+escape_char_for_json(char const &c, bool &parsing_key)
+{
+  switch (c) {
+  case '\'':
+    return {"\\\'"};
+  case '"':
+    return {"\\\""};
+  case '\\':
+    return {"\\\\"};
+  case '\b':
+    return {"\\b"};
+  case '\f':
+    return {"\\f"};
+  case '\t':
+    return {"\\t"};
+
+  // Special header reformating
+  case '\r':
+    return {""};
+  case '\n':
+    parsing_key = true;
+    return {"',\r\n\t'"}; // replace new line with pair delemiter
+  case ':':
+    if (parsing_key) {
+      return {"' : "}; // replace colon after keywith quote + colon
+    }
+    return {":"};
+  case ' ':
+    if (parsing_key) {
+      parsing_key = false;
+      return {"'"}; // replace first space after the key to be a quote
+    }
+    return {" "};
+  default:
+    return {&c, 1};
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////
+// Dump a header on stderr, useful together with TSDebug().
+void
+print_headers(TSHttpTxn txn, TSMBuffer bufp, TSMLoc hdr_loc, std::stringstream &ss)
+{
+  TSIOBuffer output_buffer;
+  TSIOBufferReader reader;
+  TSIOBufferBlock block;
+  const char *block_start;
+  int64_t block_avail;
+  bool parsing_key    = true;
+  size_t print_rewind = ss.str().length();
+  output_buffer       = TSIOBufferCreate();
+  reader              = TSIOBufferReaderAlloc(output_buffer);
+
+  ss << "\t'";
+  /* This will print just MIMEFields and not the http request line */
+  TSMimeHdrPrint(bufp, hdr_loc, output_buffer);
```

Comment:
Why are you printing out all the header and then parsing them instead of using the TS APIs to loop over all the headers?